Things to address in simulation:

- compare with background selection to neutral
- variation in recombination rate - still finds background selection?

Things to describe in simulation results:

- number of PCs chosen
- what weighting fixes
- what trace normalization fixes 
    (mutation density - choose windows with different numbers of snps)

Simulations:

All on a 8x8 grid (square to induce PC switching),
subpopulation size 100, for 32K generations.
Neutral mutation rate 1e-5 (so divergence will be 0.1).

Sexual, one chromosome of length 1 morgan.

- Neutral, flat recombination.
- Neutral, with recombination rate increasing linearly from 0 to 2.
- Neutral, with recombination hot spots.
- Recombination as above, with selection against mutants with s=.01 at 1,000 uniformly distributed sites.


Other points to discuss:

- relate to PCAdmix, explain difference in introduction
- show different window choice results we reference
- clarify we checked for variation in missingness, provide plot of MDS against missingness
- could neutrality + admixture explain drosophila and medicago? reference recent Neanderthal papers.
- make simple description of "weighted PCA" and decide what to do about it 
- clarify that block jackknife is approximately OK even though adjacent blocks are correlated
- clarify why didn't use recomb rates in medicago

To do on the data:

- Re-run with k=5, and add plots to supplement
- Include plots for different size windows
- Include plots for Medicago with window size in bp
- Include a plot about missing data 
    (maybe density of missing sites or correlation of missingness with genome-wide PC1?)
- PC switching


    * PC1/2 switching: make more clear in the paper that we take care of this with the math
    * PC2/3 switching: re-run with more PCs to check things don't change

